src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp

```diff
@@ -1934,6 +1934,7 @@ CSubqueryHandler::FRemoveExistentialSubquery(
 		// for existential subqueries, any column produced by inner expression
 		// can be used to check for empty answers; we use first column for that
 		CColRef *colref = pexprInner->DeriveOutputColumns()->PcrFirst();
+		colref->MarkAsUsed();
```

Comment:
Does the same changes are needed [here](https://github.com/arenadata/gpdb/blob/048a7eeefd162760ae9526de2c0ac37bdb86b9dc/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1304)?

Reply:
> Does the same changes are needed [here](https://github.com/arenadata/gpdb/blob/048a7eeefd162760ae9526de2c0ac37bdb86b9dc/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1304)?

It seems, yes, I added it, but I could not find a request for verification, because in the test request, it gets here already after the forced usage marking.

Reply:
How about another approach?

Reply:
> How about another approach?

Any comment about why you decided to change your mind and why this solution is better?

Reply:
I didn't change my mind, I just made the solution more general by marking the attribute as used before putting it in the array.

Reply:
> I didn't change my mind

You changed your mind about the place of the fix. Why the new place is better and why it's ok to change the function which is used in many different places? Tell me your logic.

Reply:
> Why the new place is better

I think the fix at this point might cover other cases not yet found.

Reply:
> Despite it looks very logical, we can't change the code without the understanding of what it changes.

Then we back to the quoted statement.

Reply:
We hit CSubqueryHandler.cpp:1936 from CXformSubqueryUnnest:185 and then hit CSubqueryHandler.cpp:1304 from CXformSubqueryUnnest:186 just because of the last fEnforceCorrelatedApply argument.
Therefore, it is enough to add a usage mark after CSubqueryHandler.cpp:1936 and then the used attribute will always come to CSubqueryHandler.cpp:1304.

Reply:
So I can only keep the first commit in the pull request and delete the rest.

Reply:
> we always meet CSubqueryHandler.cpp:1304 after CSubqueryHandler.cpp:1936 for the all queries we know?

No, for example:

[CSubqueryHandler::FCreateCorrelatedApplyForQuantifiedSubquery](https://github.com/arenadata/gpdb/blob/3756bbba81d9e01ee5a8b4d2238734c47fa0c43e/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1227-L1229) -> [CSubqueryHandler::FRemoveExistsSubquery](https://github.com/arenadata/gpdb/blob/3756bbba81d9e01ee5a8b4d2238734c47fa0c43e/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1981-L1986) -> [CSubqueryHandler::FCreateCorrelatedApplyForExistentialSubquery](https://github.com/arenadata/gpdb/blob/3756bbba81d9e01ee5a8b4d2238734c47fa0c43e/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1304)
while
[CXformUtils::SubqueryAnyToAgg](https://github.com/arenadata/gpdb/blob/1c2c49c78e9de67ec2abe5ffe12759a4ec550f23/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp#L842-L845) -> [CSubqueryHandler::FRemoveExistsSubquery](https://github.com/arenadata/gpdb/blob/3756bbba81d9e01ee5a8b4d2238734c47fa0c43e/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1988-L1990) -> [CSubqueryHandler::FRemoveExistentialSubquery](https://github.com/arenadata/gpdb/blob/3756bbba81d9e01ee5a8b4d2238734c47fa0c43e/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp#L1936)

Reply:
What query can hit only CSubqueryHandler.cpp:1304? Did you test such query trying to modify and break it?

Reply:
with diff
```diff
diff --git a/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp b/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
index 2ce541a472..bd454265c0 100644
--- a/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
+++ b/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
@@ -58,6 +58,8 @@
 
 using namespace gpopt;
 
+static int hit = 0;
+
 #ifdef GPOS_DEBUG
 //---------------------------------------------------------------------------
 //	@function:
@@ -1302,6 +1304,8 @@ CSubqueryHandler::FCreateCorrelatedApplyForExistentialSubquery(
 	// for existential subqueries, any column produced by inner expression
 	// can be used to check for empty answers; we use first column for that
 	CColRef *colref = pexprInner->DeriveOutputColumns()->PcrFirst();
+	hit--;
+	GPOS_RTL_ASSERT(0 == hit);
 
 	pexprInner->AddRef();
 	if (EsqctxtFilter == esqctxt)
@@ -1934,6 +1938,7 @@ CSubqueryHandler::FRemoveExistentialSubquery(
 		// for existential subqueries, any column produced by inner expression
 		// can be used to check for empty answers; we use first column for that
 		CColRef *colref = pexprInner->DeriveOutputColumns()->PcrFirst();
+		hit++;
 
 		if (COperator::EopScalarSubqueryExists == op_id)
 		{
```
the string CSubqueryHandler.cpp:1304 is reached before the string CSubqueryHandler.cpp:1936 on query from unit test DynamicIndexGet-OuterRefs.mdp

Reply:
> What query can hit only CSubqueryHandler.cpp:1304?

```diff
diff --git a/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp b/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
index 2ce541a472..bd454265c0 100644
--- a/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
+++ b/src/backend/gporca/libgpopt/src/xforms/CSubqueryHandler.cpp
@@ -58,6 +58,8 @@
 
 using namespace gpopt;
 
+static int hit = 0;
+
 #ifdef GPOS_DEBUG
 //---------------------------------------------------------------------------
 //	@function:
@@ -1302,6 +1304,8 @@ CSubqueryHandler::FCreateCorrelatedApplyForExistentialSubquery(
 	// for existential subqueries, any column produced by inner expression
 	// can be used to check for empty answers; we use first column for that
 	CColRef *colref = pexprInner->DeriveOutputColumns()->PcrFirst();
+	hit--;
+	GPOS_RTL_ASSERT(0 == hit);
 
 	pexprInner->AddRef();
 	if (EsqctxtFilter == esqctxt)
@@ -1934,6 +1938,7 @@ CSubqueryHandler::FRemoveExistentialSubquery(
 		// for existential subqueries, any column produced by inner expression
 		// can be used to check for empty answers; we use first column for that
 		CColRef *colref = pexprInner->DeriveOutputColumns()->PcrFirst();
+		hit++;
 
 		if (COperator::EopScalarSubqueryExists == op_id)
 		{
```
shows that the string CSubqueryHandler.cpp:1304 is reached before the string CSubqueryHandler.cpp:1936 on query from unit test DynamicIndexGet-OuterRefs.mdp

Reply:
> Did you test such query trying to modify and break it?

Yes, I changed the queries from all three tests, got an unused attribute on CSubqueryHandler.cpp:1304, but the CTranslatorExprToDXL.cpp:7580 assert did not work in all three cases, because reqd_prop_plan did not contain this attribute.

Reply:
Thus, another solution suggests itself: remove unused attributes from reqd_prop_plan.

Reply:
I suggest another approach, which gets first used column for exist subquery